Dockerfile

```diff
@@ -23,7 +23,8 @@ RUN --mount=type=cache,target=/var/cache/apt \
     rm /usr/bin/python3 &&\
     ln -s /usr/bin/python3.10 /usr/bin/python3 &&\
     # Create a fake system Python pointing at venv python
-    echo 'exec /opt/venv/bin/python3.10 "$@"' > /usr/bin/python &&\
+    printf "#!/bin/bash\n/opt/venv/bin/python3.10 \"\$@\"" > /usr/bin/python &&\
```

Comment:
Could you tell me more about what we're accomplishing with `printf` here? I can see that we're writing two lines to a file, but why `printf` and not `echo`?

I appreciate that you didn't last edit this line (@lucyb did 🙂), but could you also help me understand why we're creating a fake system Python and not activating the venv? By not activating the venv, the user doesn't get access to useful tools in `/opt/venv/bin`, such as `ipython` and `pytest`.

Reply:
a further Google tells me that `echo -e` supports newline characters, I've always previously reached for `printf` in that circumstance.

As to why - this stemmed from a chat with Dave and Simon about the possibility of redirecting the system python to one running inside (one of) the OpenSAFELY Python docker container(s), this is scaffolding to make it easier to make that switch in the future. It also lets us point to the venv python before it's activated which fixed another problem I can't quite remember right now.


re: venv activation - the venv is activated by vs code when the codespace/devcontainer starts up. Which venv VSCode chooses to activate is configured by the Python default interpreter setting. This worked fine when we had vs code settings in a vs code settings file in the research template, but became flaky when these settings were moved to the devcontainer.json. By faking the system python and removing the interpreter path setting, vs code reliably activates the venv we want it to at startup

Reply:
Thanks, @Jongmassey.

To make the intent of this commit clear, could you use `echo -e`? Diffs are super-informative; the dark green background of `-e` would be just enough to communicate "We want newlines now, but we didn't before".

It's a shame there isn't a comment about the scaffolding. Could you add one?

I can't see where VS Code activates the venv in [`.devcontainer/devcontainer.json`](https://github.com/opensafely/research-template/blob/improve-devcontainers/.devcontainer/devcontainer.json). Is doing so a feature of `ms-python.python`?

Whatever should be activating the venv, the venv isn't being activated. We can verify this by typing `ipython` in the terminal. The result is `command not found`. I also see that the `ehrql` imports in the dataset definition are raising syntax errors. Have I misunderstood? If not, I'll create an issue.

Reply:
https://github.com/opensafely/research-template/blob/improve-devcontainers/.devcontainer/devcontainer.json#L42-L43

>                "python.terminal.activateEnvironment": true,
>                "python.terminal.activateEnvInCurrentTerminal": true,

Reply:
OK so activation of the venv in the terminal isn't actually working with this despite me thinking that it did when I tried it locally. I'll poke at it some more until it does.

Reply:
Thank you 🙂

Reply:
Actually, @Jongmassey, let's deal with the virtual environment issue separately:

* opensafely-core/research-template-docker#19
* opensafely-core/research-template-docker#20

We can triage these issues next week.